### PR TITLE
README: More detailed description of DBSP.

### PR DESCRIPTION
**Pull request by ryzhyk**

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -7,23 +7,57 @@
...
+1. **Full SQL support and more.**  While SQL is just the first of potentially
+many DBSP frontends, it offers a reference point to chatacterize the
+expressiveness of the engine.  Our goal is to support the complete SQL syntax
+and semantics, including joins and aggregates, correlated subqueries, window
+functions, complex data types, time series operators, UDFs, etc.  Beyond
+standard SQL, DBSP supports recursive queries, which arise for instance in graph
+analytics problems.
 
-## Overview
+1. **Scalability in multple dimensions.**  The engine scales with the number and
```
Comment (mbudiu-vmw) [inline README.md:31]: multiple

File: README.md [hunk 2]
```diff
@@ -7,23 +7,57 @@
...
-world that increasingly relies on real-time alerts, recommendations, control, etc.
+We believe that software engineers and data scientists who build streaming data
+pipelines should not be exposed to this complexity.  They should be able to
+express their computations as declarative queries and use a streaming engine to
+evaluate these queries correctly and efficiently.  DBSP aims to be such an
+engine.  To this end we set the following high-level objectives:
 
-DBSP intends to change the situation by offering a simple promise: to **execute
-any query workload that can run in batch mode in streaming mode**.
+1. **Full SQL support and more.**  While SQL is just the first of potentially
```
Comment (mbudiu-vmw) [inline README.md:23]: technically there is nothing about SQL in this repository.
Perhaps more precise it is to day "support for mainstream languages, including full standard SQL"

File: README.md [hunk 3]
```diff
@@ -7,23 +7,57 @@
...
+functions, complex data types, time series operators, UDFs, etc.  Beyond
+standard SQL, DBSP supports recursive queries, which arise for instance in graph
+analytics problems.
 
-## Overview
+1. **Scalability in multple dimensions.**  The engine scales with the number and
+complexity of queries, streaming data rate and the amount of state the system
+maintains in order to process the queries.
+
+1. **Performance out of the box.**  The user should be able to focus on the
```
Comment (mbudiu-vmw) [inline README.md:35]: This is a tricky one - the query plan you start from will be as important if not more important than the incrementalization. you can always write stupid plans.

File: README.md [hunk 4]
```diff
@@ -7,23 +7,57 @@
...
+business logic of their application, leaving it to the system to evaluate this
+logic efficiently.
+
+## Theory
+
+The above objectives can only be achieved by building on a solid mathematical
+foundation.  The formal model that underpins our system, also called DBSP, is
+described in the accompanying paper:
+
+- [Budiu, McSherry, Ryzhyk, Tannen. DBSP: Automatic Incremental View Maintenance
```
Comment (mbudiu-vmw) [inline README.md:45]: you now have the latex sources in the same repo, and I should check-in the long pdf as well soon.

File: README.md [hunk 5]
```diff
@@ -7,23 +7,57 @@
...
+  for Rich Query Languages](https://arxiv.org/abs/2203.16684)
+
+The model provides two things:
+
+1. **Semantics.** DBSP defines a formal language of streaming operators and
+queries built out of these operators, and precisely specifies how these queries
+must transform input streams to output streams.
+
+1. **Algorithm.** DBSP also gives an algorithm that takes an arbitrary query and
```
Comment (mbudiu-vmw) [inline README.md:54]: the funny thing is that the algorithm is not used anywhere by the code in this repository

Comment (mbudiu-vmw) [reply README.md:54]: you should specify that this is this repository is the platform on top of which query engines can be implemented

File: README.md [hunk 6]
```diff
@@ -59,19 +93,31 @@ In addition, DBSP supports **windowing operators** that group time series data
...
-  for Rich Query Languages](https://arxiv.org/abs/2203.16684)
+<p align="center">
+  <img src="architecture.png" width="400" alt="DBSP architecture">
+</p>
+
+The DBSP core engine is written in Rust and provides a Rust API for building
+data-parallel dataflow programs by instantiating and connecting streaming
+operators.  Developers can use this API directly to implement complex
+streaming queries.  We are also developing a
+[compiler from SQL to DBSP](https://github.com/vmware/sql-to-dbsp-compiler) that will
```
Comment (mbudiu-vmw) [inline README.md:110]: use the present tense

File: README.md [hunk 7]
```diff
@@ -59,19 +93,31 @@ In addition, DBSP supports **windowing operators** that group time series data
...
+like Python and Scala.
 
-## Front-end languages
+At runtime, DBSP can consume inputs from and send outputs to
+event streams, e.g., Kafka, databases, e.g., Postgres, and data warehouses,
+e.g., Snowflake.
 
-In case you don't like Rust, a compiler from SQL to DBSP is being
-developed: <https://github.com/vmware/sql-to-dbsp-compiler>
+The distributed runtime extends DBSP's data-parallel execution model to
```
Comment (mbudiu-vmw) [inline README.md:119]: Here you can use the future tense

File: README.md [hunk 8]
```diff
@@ -59,19 +93,31 @@ In addition, DBSP supports **windowing operators** that group time series data
...
-Delivering this functionality with strong performance and correctness guarantees
-requires building on a solid foundation.  The theory behind DBSP is described in
-the accompanying paper:
+The following diagram shows the architecture of the DBSP platform.  Solid
+blocks indicate components that we are currently working on; white blocks with
+dashed borders are on our TODO list.
 
-- [Budiu, McSherry, Ryzhyk, Tannen. DBSP: Automatic Incremental View Maintenance
-  for Rich Query Languages](https://arxiv.org/abs/2203.16684)
+<p align="center">
```
Comment (mbudiu-vmw) [inline README.md:102]: the diagram is too specific about python and scala. Perhaps DDlog will come first?
I would put "other query languages"

Comment (mbudiu-vmw) [reply README.md:102]: same for the adapters. Instead of postgres you should perhaps put ODBC and CDC

**Review by mbudiu-vmw — APPROVED** [review]
The sections you added are more advertisment than manual.

File: README.md (hunk 2 above)
Comment (ryzhyk) [reply README.md:23]: I already have a pretty detailed remark saying that SQL is the first of several.  It's true that this README is for the whole project, not just this repo though.  I think this is ok.

**Pull request merged**